f5_openstack_agent/lbaasv2/drivers/bigip/icontrol_driver.py

```diff
@@ -1162,61 +1155,63 @@ def get_loadbalancers_in_tenant(self, tenant_id):
+                                                  plugin_const.ERROR)
         try:
-            self.tenant_manager.assure_tenant_created(service)
+            try:
+                self.tenant_manager.assure_tenant_created(service)
+            except Exception as e:
+                LOG.error("Tenant folder creation exception: %s",
+                          e.message)
+                if lb_provisioning_status != plugin_const.PENDING_DELETE:
+                    loadbalancer['provisioning_status'] = \
+                        plugin_const.ERROR
+                raise e
+
             LOG.debug("    _assure_tenant_created took %.5f secs" %
                       (time() - start_time))
 
             traffic_group = self.service_to_traffic_group(service)
-            service['loadbalancer']['traffic_group'] = traffic_group
-
-            # This loop will only run once.  Using while as a control-flow
-            # mechanism to flatten out the code by allowing breaks.
-            while (self.network_builder):
-
-                if not self.service_adapter.vip_on_common_network(service) \
-                    and not self.disconnected_service.\
-                        is_service_connected(service):
-                    if self.disconnected_service_polling.enabled:
-                        # Hierarchical port-binding mode:
-                        # Skip network setup if the service is not connected.
-                        break
-                    else:
-                        LOG.error("Misconfiguration: Segmentation ID is "
-                                  "missing from the service definition. "
-                                  "Please check the setting for "
-                                  "f5_network_segment_physical_network in "
-                                  "f5-openstack-agent.ini in case neutron "
-                                  "is operating in Hierarhical Port Binding "
-                                  "mode.")
-                        service['loadbalancer']['provisioning_status'] = \
-                            plugin_const.ERROR
-                        raise f5ex.MissingNetwork("Missing segmentation id")
+            loadbalancer['traffic_group'] = traffic_group
 
+            if self.network_builder:
                 start_time = time()
                 try:
                     self.network_builder.prep_service_networking(
                         service, traffic_group)
-                except Exception as exc:
-                    LOG.error("Exception: icontrol_driver: %s", exc.message)
-                    service['loadbalancer']['provisioning_status'] = \
-                        plugin_const.ERROR
-                    raise
-
-                if time() - start_time > .001:
-                    LOG.debug("    _prep_service_networking "
-                              "took %.5f secs" % (time() - start_time))
-                break
+                except f5ex.NetworkNotReady as error:
```

Comment:
If the loadbalancer VIP network is not ready, The loadbalancer is in the 'PENDING_CREATE' state.  We return and leave it for the agent_manager to query for "PENDING" loadbalancers.  Once the network is complete, we can get past this error.

IF the member network is not ready, the loadbalancer is in the PENDING_UPDATE state, the member is in PENDING_CREATE.  In this case we let the agent_manager query for PENDING loadbalancers, PENDING_CREATE or PENDING_UPDATE.  We don't really need to know anything about the member since the loadbalancer will be locked until we can complete the member create.  Since we are looking for loadbalancers in PENDING_UPDATE state too, we will pull the member definition over eventually.  If not, the loadbalancer will be put in ERROR state.  The member should also be put into error state, but it doesn't.  This is an issue that I don't address.  We should either, make the member in error state or delete it.